logview/grablogs.conf.4
'\" te
.\" gnome-system-log.1 - GNOME logfile viewer
.\" Copyright 1989 AT
.\"|[amp   ]
.\"|T
.\" Copyright 2006 Lin Ma
.TH grablogs.conf 4 "13 Oct 2006" "gnome-utils 2.16.0" "File Formats"
.SH "NAME"
grablogs.conf \- grablogs configuration for libgrablogs.so of the plugins of
 gnome-system-log
file
.SH "SYNOPSIS"
.PP
.nf
/usr/lib/gnome-system-log/plugins/grablogs.conf
.fi
.SH "DESCRIPTION"
.PP
The libgrablogs.so is a plugin for gnome-system-log(1), it colloct the log 
files from the system as many as possible. grablogs.conf is a configuration 
file that contains a set of lines mixed with sh(1) syntax codes and individual
 log files. libgrablogs.so will read the file try to get a log files list for
 gnome-system-log(1). Users can copy the file into $HOME/.gnome2/gnome-system-log/plugins/`uname -p` to overwrite the system default one.
.PP
The grablogs.conf file contains the following configuration categories:
.RS
.TP
.B [configs]
Each line under this category is interpreted as a config file of
System. The plugin will open the config file and try to find all
system paths of the logs.
.TP
.B [commands]
Each line under this category is interpreted as a shell command and
will be execute through a pipe. And each line of the output of the
command will be interpreted as a log path.
.TP
.B [logs]
Each line under this category is interpreted as a log path.
.RE
.SH "FILES"
.TP
.I /usr/lib/gnome-system-log/plugins/grablogs.conf
The system default configuration file for the plugin libgrablogs.so
.TP
.I "$HOME/.gnome2/gnome\-system\-log/plugins/`uname \-p`/grablogs.conf"
The user specific configuration file for the plugin libgrablogs.so
.SH "EMAMPLE"
.br
.B [configs]
.br
.IR /etc/syslog.conf
.br
.B [commands]
.br
.IR "for i in `svcs -aH -o FMRI | grep -v lrc `; do svcprop -p restarter/logfile $i 2>/dev/null || svcprop -q -p restarter/alt_logfile $i 2>/dev/null ; done"
.br
.B [logs]
.br
.IR /var/log/messages
.br
.IR /var/log/secure
.br
.IR /var/log/maillog
.br
.IR /var/log/cron
.br
.IR /var/log/Xorg.0.log
.br
.IR /var/log/XFree86.0.log
.br
.IR /var/log/auth.log
.br
.IR /var/log/cups/error_log
.SH "SEE ALSO"
.PP
.BR gnome-system-log (1),
.BR pipelog.conf (1)
.PP


NEWS
===================================================================
GNOME Utilities 2.18.1 (14th April 2007) - "Sound of Silver"

    * Second stable release of the 2.18 cycle

Baobab
Dictionary
Floppy
Screenshot
Search Tool

    * Nothing

System Log Viewer

    * Fix bug #364106 (and countless duplicates) - crash in System
      Log: string_get_date_string() at logrtns.c:136. [Jan Arne Petersen]

Translations Updates By:

    * Peter Bach (da)
    * Pema Geyleg (dz)
    * Claudio Saavedra (es)
    * Inaki Larranaga Murgoitio (eu)
    * Ignacio Casal Quinteiro (gl)
    * Leonardo Ferreira Fontenelle (pt_BR)
    * Baris Cicek (tr)

===================================================================
GNOME Utilities 2.18.0 (11th March 2007) - "Don't Get Lost in Heaven"

    * First stable release of the 2.18 cycle

Baobab

    * Fix bug #411867 - baobab crashed with SIGSEGV in
      pango_fc_font_lock_face()
    * Resort when toggling "allocated space".

Dictionary

    * Nothing

Floppy

    * Nothing

Screenshot

    * Fix bug #413019 - The effects in theinteractive dialog are not
      fully translated
    * Fix bug #411050 - With some window name, the screenshot tool
      will propose to replace ~/Desktop with the image

Search Tool

    * Nothing

System Log Viewer

    * Nothing

Translations Updated By:

    Djihed Afifi (ar)
    Alexander Shopov (bg)
    Josep Puigdemont (ca)
    Hendrik Richter (de)
    norbu (dz)
    David Lodge (en_GB)
    Ivar Smolin (et)
    Ilkka Tuohela (fi)
    Claude Paroz (fr)
    Ankit Patel (gu)
    Gabor Kelemen (hu)
    Andrea Zagli (it)
    Takeshi AIHANA (ja)
    Changwoo Ryu (ko)
    Erdal Ronahi (ku)
    Gintautas Miliauskas (lt)
    Jovan Naumovski (mk)
    Kjartan Maraas (nb)
    wadim dziedzic (pl)
    Duarte Loreto (pt)
    Leonardo Ferreira Fontenelle (pt_BR)
    Leonid Kanter (ru)
    Daniel Nylander (sv)
    Theppitak Karoonboonyanan (th)
    Maxim Dziumanenko (uk)
    Clytie Siddall (vi)
    Funda Wang (zh_CN)
    Woodman Tuen (zh_HK, zh_TW)

===================================================================
GNOME Utilities 2.17.92 (12th February 2007) - "You'll not feel the drowning"

    * Fifth unstable release of the 2.17 cycle.
    * Fix bug #404236 - GFloppy should not be build on systems
      without a floppy drive.

Baobab

    * Nothing

Dictionary

    * Fix bug #376937 - Fix build with gcc 2.95

Floppy
    
    * Nothing

Screenshot

    * Clean up the interactive dialog code.
    * Hide the interactive dialog as soon as possible and avoid
      errors when grabbing a window without a delay.
    * Use a better interpolation algorithm for the border
      effect.
    * Fix bug #400200 - Do not leak the GnomeProgram instance
      when quitting. [Christian Persch]
    * Fix part of bug #400203 - HIG fixes for the interactive
      dialog. [Christian Persch]

Search Tool

    * Fix bug #397945 - Cursor goes away from results list when
      deleting a file.
    * Fix bug #395381 - Cannot delete filenames containing %21
      from the search window.
    * Fix bug #404158 - Do not overwrite files in the trash folder.
      Use same file renaming logic as nautilus.

System Log Viewer

    * Nothing.

Translations Updated By:

    * Khaled Hosny (ar)
    * Josep Puigdemont i Casamajó (ca)
    * David Lodge (en_GB)
    * Priit Laes,  Ivar Smolin (et)
    * Ilkka Tuohela (fi)
    * Stéphane Raimbault, Jonathan Ernst (fr)
    * Changwoo Ryu (ko)
    * Kjartan Maraas (nb)
    * Leonardo Ferreira Fontenelle, Luiz Fernando S. Armesto (pt_BR)
    * Daniel Nylander (sv)
    * Theppitak Karoonboonyanan (th)
    * Nguyễn Thái Ngọc Duy (vi)

===================================================================
GNOME Utilities 2.17.91 (23rd January 2007) - "The Small Print"

    * Fourth unstable release of the 2.17 cycle.
    * Fix an assertion failure in the Screenshot utility.

Baobab
Dictionary
Floppy
Search Tool
System Log Viewer
    
    * Nothing

Screenshot

    * Fix bug #399859 - crash in Take Screenshot

Translations Updates By:

    * Stéphane Raimbault, Robert-André Mauchin (fr)
    * Daniel Nylander (sv)
    * Theppitak Karoonboonyanan (th)

===================================================================
GNOME Utilities 2.17.90 (22nd January 2007) - "Small Stakes"

    * Third unstable release of the 2.17 cycle.

Baobab

    * Fix an assertion failure. [Paolo Borelli]

Dictionary

    * Fix bug #395837 - Dictionary simply doesn't show a result and
      no error message
    * Fix bug #395125 - "Unable to find dictionary source" when
      launching the dictionary

Floppy

    * Nothing.

Screenshot

    * Add the available effects to the interactive dialog.
    * Show the screenshot icon. [Dennis Cranston]

Search Tool

    * Nothing.

System Log Viewer

    * Fix bug #399544 - Typo in message string

Translation Updates By:

    * Djihed Afifi (ar)
    * Hendrik Richter (de)
    * David Lodge (en_GB)
    * Daniel Nylander (sv)
    * Theppitak Karoonboonyanan (th)

===================================================================
GNOME Utilities 2.17.1 (8th January 2007) - "Cast no shadow"

    * Second unstable release of the 2.17 cycle.

Baobab

    * Small clean ups.

Dictionary

    * Fix bug #166015 - Lookup selected text automatically
    * Fix bug #342851 - No longer renders cross references as hyperlinks
    * Fix bug #333132 - Allow editing of dictionary sources
    * Fix bug #352141 - Problems in the View menu
    * Fix bug #351601 - Add dict advanced settings not working
    * Fix bug #377296 - The gdk-doc build is broken.
    * Fix bug #372386 - Add Thai dictionary source
    * Fix bug #348656 - Poor discoverability of the look up process
    * Make --match=word command line argument work again

Floppy

    * Nothing.

Screenshot

    * Fix bug #161525 - Screenshot application complains that it cannot
      save the screenshot file on *BSD.
    * Fix bug #362903 - gnome-panel-screenshot asks for password, and

logview/plugins/grablogs.c
/* -*- Mode: C; indent-tabs-mode: t; c-basic-offset: 8; tab-width: 8 -*- */
/*
 * Copyright (C) 2006 Lin Ma <Lin.Ma@sun.com>
 */
 
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <string.h>
#include <glib.h>
#include <stdio.h>
#include <signal.h>
#include <glib.h>
#include <locale.h>
#include "logview-plugin.h"
#include "logview-iface-collector.h"
#include "logview-debug.h"
#include "logview-log.h"
#include "misc.h"

#define PLUGIN_GRABLOGS(obj) \
	(G_TYPE_CHECK_INSTANCE_CAST ((obj), PLUGIN_TYPE_GRABLOGS, PluginGrabLogs))
#define PLUGIN_GRABLOGS_CLASS(klass) \
	(G_TYPE_CHECK_CLASS_CAST ((klass), PLUGIN_TYPE_GRABLOGS, PluginGrabLogsClass))
#define GRABLOGS_IS_PLUGIN(obj) \
	(G_TYPE_CHECK_INSTANCE_TYPE ((obj), PLUGIN_TYPE_GRABLOGS))
#define GRABLOGS_IS_PLUGIN_CLASS(klass) \
	(G_TYPE_CHECK_CLASS_TYPE ((klass), PLUGIN_TYPE_GRABLOGS))
#define PLUGIN_GRABLOGS_GET_CLASS(obj) \
	(G_TYPE_INSTANCE_GET_CLASS ((obj), PLUGIN_TYPE_GRABLOGS, PluginGrabLogsClass))

#undef PLUGIN_PRIO
#define PLUGIN_PRIO	PLUG_PRIO_BASE
#undef PLUGIN_DESC
#define PLUGIN_DESC	"For colloct log files from shell commands"

#ifdef ON_SUN_OS
#define FILE_CMD "file"
#else
#define FILE_CMD "file"
#endif
#define GRABLOGS_CONF	"grablogs.conf"
#define CAT_CMD		"[commands]"
#define CAT_LOG		"[logs]"
#define CAT_CONF	"[configs]"


typedef GSList* (*Parse_All) (FILE*);
typedef GSList* (*Parse_Line) (FILE*, Parse_All);
typedef struct _PluginGrabLogs PluginGrabLogs;
typedef struct _PluginGrabLogsClass PluginGrabLogsClass;

struct _PluginGrabLogs {
	LogviewPlugin parent;
};

struct _PluginGrabLogsClass {
	LogviewPluginClass parent;
};

/* class implement */
static void iface_init (gpointer g_iface, gpointer iface_data);
static void collector_init (gpointer g_iface, gpointer iface_data);

static struct sigaction old_sa;
static gchar* category;

static void
grablogs_class_init (gpointer g_class, gpointer g_class_data)
{
	LV_MARK;
}

static void
grablogs_instance_init (GTypeInstance *instance, gpointer g_class_data)
{
	LV_MARK;
}

static GType
grablogs_get_type (GTypeModule* module)
{
	static GType type = 0;
	LV_MARK;
	if (type == 0) {
		static const GTypeInfo info = {
			/* You fill this structure. */
			sizeof (PluginGrabLogsClass),
			NULL,   /* base_init */
			NULL,   /* base_finalize */
			grablogs_class_init,   /* class_init */
			NULL,   /* class_finalize */
			NULL,   /* class_data */
			sizeof (PluginGrabLogs),
			0,      /* n_preallocs */
			grablogs_instance_init    /* instance_init */
		};
		static const GInterfaceInfo iface_info = {
			iface_init,	/* interface_init */
			NULL,		/* interface_finalize */
			NULL		/* interface_data */
		};
		static const GInterfaceInfo collector_info = {
			collector_init,	/* interface_init */
			NULL,		/* interface_finalize */
			NULL		/* interface_data */
		};
		type = g_type_module_register_type (module,
						    LOGVIEW_TYPE_PLUGIN,
						    "GrablogsType",
						    &info,
						    0);
		g_type_module_add_interface (module, type, LOGVIEW_TYPE_IFACE,
			&iface_info);
		g_type_module_add_interface (module, type, LOGVIEW_TYPE_IFACE_COLLECTOR,
			&collector_info);
	}
	return type;
}

static GSList*
get_logs (const gchar* logcfg, Parse_All cp)
{
	GSList* loglist = NULL;
	FILE* handle;
	int (*lclose) (FILE*) = NULL;
	
	LV_MARK;
	/*
	 * lines in [command] will be run in popen
	 */
	if (category && g_ascii_strcasecmp(g_strstrip(category),
					   CAT_CMD) == 0) {
		LV_INFO ("[grablogs-PIPE-open] %s", logcfg);
		handle = security_popen (logcfg, "r");
		lclose = security_pclose;
	}
	else if (local_file_exist (logcfg)
		 && local_file_can_executed (logcfg)) {
		LV_INFO ("[grablogs-FILE-open] %s", logcfg);
		handle = fopen (logcfg, "r");
		lclose = fclose;
	}
	else
		goto get_logs_err;

	if (handle != NULL) {
		LV_INFO ("[grablogs-open] %s", logcfg);
		loglist = cp (handle);
		if (lclose (handle) == -1)
			goto get_logs_err;
		LV_INFO ("[grablogs-close] %s", logcfg);
	}
	else {
get_logs_err:
		LV_INFO_EE ("[grablogs handles]: \"%s\", msg: %s",
			    logcfg, error_system_string ());
		return NULL;
	}
	return loglist;
}

static gchar*
get_line (FILE *handle)
{
	glong inc = BUFSIZ, size = inc;
	glong p = -1;
	gchar *buf = (gchar *) g_malloc (size * sizeof (gchar));
	g_assert (handle);
	g_assert (buf);
	do {
		g_assert (p < size);
		if (++p == size) {
			size += inc;
			buf = (gchar *) g_realloc (buf, size * sizeof(gchar));
			g_assert (buf);
		}
		buf[p] = (gchar)getc (handle);
	} while (buf[p] != '\n' && buf[p] != EOF);
	if (buf[p] != EOF) {
		buf[p] = '\0';
	} else if (feof (handle) == 0 && p > 0) {
		buf[p] = '\0';
	} else if (feof (handle) != 0) {
		LV_INFO_EE ("get_line error: %s", error_system_string());
		g_free (buf);
		buf = NULL;
	} else {
		g_free (buf);
		buf = NULL;
	}
	LV_ERR ("get_line %s", buf? buf:"EOFed");
	return buf;
}

/* one log per line */
static GSList*
cp_logline (FILE *handle)
{
	gchar *buf;
	GSList *logfiles = NULL;
	LV_MARK;
	while ((buf = get_line(handle)) != NULL) {
		if (g_str_has_prefix (g_strchug (buf), "#")) {
			g_free (buf);
			continue;
		}
		if (g_str_has_prefix(g_strstrip(buf), "[")) {
			g_free (category);
			category = buf;
			break;
		}
		if (strlen (buf) != 0)
			logfiles = g_slist_append (logfiles, g_strdup (buf));
		g_free (buf);
	}
	return logfiles;
}


static GSList*
cp_logconf (FILE *handle)
{
	gchar *buf, *p;
	GSList *logfiles = NULL;
	LV_MARK;
	while ((buf = get_line(handle)) != NULL) {
		gchar **list;
		gint i;
		for (p = buf; g_ascii_isspace(*p); ++p);
		if (*p == '\0' || *p == '#' || *p == '\n') {
			g_free (buf);
			continue;
		}
		list = g_strsplit_set (p, ", -\t()\n'`", 0);
		for (i = 0; list[i]; ++i) {
			if (*list[i] == '/' &&
			    g_slist_find_custom(logfiles, list[i],
						(GCompareFunc)g_ascii_strcasecmp) == NULL) {
				logfiles = g_slist_append (logfiles,
							   g_strdup(list[i]));
			}
		}
		g_strfreev(list);
		g_free (buf);
	}
	return logfiles;
}

/* one shell or command per line
 * run each line
 * one log per line of output
 */
static GSList*
cp_shell_per_line (FILE *handle)
{
	gchar *buf = NULL;
	GSList *all_logs = NULL;
	LV_MARK;
	while ((buf = get_line(handle)) != NULL) {
		if (g_str_has_prefix (g_strchug (buf), "#")) {
			g_free (buf);
			continue;
		}
		if (g_str_has_prefix(g_strchug (buf), "[")) {
			g_free (category);
			category = buf;
			break;
		}
		if (strlen (buf) != 0) {
			GSList *idx;
			GSList *logfiles = NULL;
			logfiles = get_logs (buf, cp_logline);
			for (idx = logfiles; idx; idx=g_slist_next(idx)) {
				if (g_slist_find_custom (all_logs, idx->data,
							 (GCompareFunc)g_ascii_strcasecmp)
				    == NULL) {
					all_logs = g_slist_append (all_logs,
								   idx->data);
				}
				else
					g_free (idx->data);
			}
			if (logfiles)
				g_slist_free (logfiles);
		}
		g_free (buf);
	}
	return all_logs;
}

static GSList*
cp_conf_per_line (FILE *handle)
{
	gchar *buf = NULL;
	GSList *all_logs = NULL;
	LV_MARK;
	while ((buf = get_line(handle)) != NULL) {
		if (g_str_has_prefix (g_strchug (buf), "#")) {
			g_free (buf);
			continue;
		}
		if (g_str_has_prefix(g_strchug (buf), "[")) {
			g_free (category);
			category = buf;
			break;
		}
		if (strlen (buf) != 0) {
			GSList *idx;
			GSList *logfiles = NULL;
			logfiles = get_logs (buf, cp_logconf);
			for (idx = logfiles; idx; idx=g_slist_next(idx)) {
				if (g_slist_find_custom (all_logs, idx->data,
							 (GCompareFunc)g_ascii_strcasecmp)
				    == NULL) {
					all_logs = g_slist_append (all_logs,
								   idx->data);
				}
				else
					g_free (idx->data);
			}
			if (logfiles)
				g_slist_free (logfiles);
		}
		g_free (buf);
	}
	return all_logs;
}

static GSList*
cp_grab_conf (FILE *handle)
{
	GSList *all_logs = NULL;
	LV_MARK;

	/* A line without any categories will be skipped */
	category = get_line(handle);
	while (feof (handle) == 0) {
		if (g_str_has_prefix (g_strchug (category), "#"))
			goto next_line;
		if (category) {
			if (g_ascii_strcasecmp(g_strstrip(category),
					       CAT_CMD) == 0) {
				LV_INFO ("grab category %s, enter", category);
				all_logs = g_slist_concat (all_logs,
							   cp_shell_per_line (handle));
			}
			else if (g_ascii_strcasecmp(g_strstrip(category),
						    CAT_LOG) == 0) {
				LV_INFO ("grab category %s, enter", category);
				all_logs = g_slist_concat (all_logs,
							   cp_logline (handle));
			}
			else if (g_ascii_strcasecmp(g_strstrip(category),
						    CAT_CONF) == 0) {
				LV_INFO ("grab category %s, enter", category);
				all_logs = g_slist_concat (all_logs,
							   cp_conf_per_line (handle));
			}
			else
				goto next_line;
		}
		else {
next_line:
			g_free (category);
			category = get_line(handle);
		}
	}
	g_free (category);
	return all_logs;
}

static GSList*
get_all_logs (PluginGrabLogs *self)
{
	GSList *plugin_paths, *i;
	GSList *all_logs = NULL;
	LV_MARK;

	plugin_paths = (GSList *) get_plugin_paths ();
	category = NULL;
	for (i = plugin_paths; i; i=g_slist_next(i)) {
		gchar* path = g_build_path (G_DIR_SEPARATOR_S,
					    i->data, GRABLOGS_CONF, NULL);
		if (local_file_exist (path)
		    && local_file_can_executed (path)) {
			GSList *logfiles, *idx;
			logfiles = get_logs (path, cp_grab_conf);
			for (idx = logfiles; idx; idx = g_slist_next(idx)) {
				if (g_slist_find_custom (all_logs, idx->data,
							 (GCompareFunc)g_ascii_strcasecmp) == NULL) {
					all_logs = g_slist_append (all_logs, idx->data);
					LV_ERR ("add %s", idx->data);
				}
				else {
					LV_ERR ("del %s", idx->data);
					g_free (idx->data);
				}
			}
			g_slist_free (logfiles);
			/* break to search the sequent path of GRABLOGS_CONF */
			g_free (path);
			break;
		} else {
			g_free (path);
		}
	}
	return all_logs;
}

static void
config_child_log (Log *log)
{
	/* Check for older versions of the log */
	gchar *log_name = NULL;
	gint i;
	g_object_get (G_OBJECT (log), "path", &log_name, NULL);
	for (i=0; i<OLD_LOG_NUM; i++) {
		gboolean flag = FALSE;
		gchar *older_name;
		Log *child;
		older_name = g_strdup_printf ("%s.%d", log_name, i);
		if (local_file_exist (older_name)) {
			flag = TRUE;
		} else {
			g_free (older_name);
			older_name = g_strdup_printf ("%s.%d.gz", log_name, i);
			if (local_file_exist (older_name)) {
				flag = TRUE;
			}
		}
		if (flag) {
			child = g_object_new (LOGVIEW_TYPE_LOG,
					      "path", older_name,
					      "protocol", file_protocol (older_name),
					      NULL);
			log_set_child (log, i, child);
		}
		g_free (older_name);
	}
	g_free (log_name);
}

static Log*
config_log_from_path (PluginGrabLogs *self, const gchar *log_path)
{
	Log *log;
	LogProtocol protocol;
	gchar *utf8_name = NULL;
	gchar *name;
	gchar *message;

	protocol = file_protocol (log_path);
	switch (protocol) {
	case UNKNOWN_LOG:
	case PLAIN_LOG:
	case GZIP_LOG:
	case PIPE_LOG:
		log = g_object_new (LOGVIEW_TYPE_LOG,
				    "path", log_path,
				    "protocol", protocol,
				    NULL);
		config_child_log (log);
		return log;
	default:;
	}
	utf8_name = locale_to_utf8 (log_path);
	name = utf8_name ? utf8_name : log_path;
	/* SUN_BRANDING */
	message = g_strdup_printf (_("Unknown file type: %d"), protocol);
	log_error_appendv ("grablogs - %s - %s", name, message);
	g_free (message);
	g_free (utf8_name);
	return NULL;
}

static void
iface_init (gpointer g_iface, gpointer iface_data)
{
	LogviewInterface *iface = (LogviewInterface*) g_iface;
	LV_MARK;
	iface->can_handle = NULL; /* dummy */
}

static void
collector_init (gpointer g_iface, gpointer iface_data)
{
	LogviewIFaceCollector *iface = (LogviewIFaceCollector*) g_iface;
	LV_MARK;
	iface->get_logs = (GSList* (*) (LogviewICollector*)) get_all_logs;
	iface->config_log_from_path = (Log* (*) (LogviewICollector*, const gchar*))config_log_from_path;
}

LOGVIEW_INIT_PLUGIN (grablogs_get_type)

